### PR TITLE
Adding b64 and json decoder and encoder functions

### LINKED ISSUE #811: scarpet should have native support for json and b64 de/serializers
maybe rename for consistency?
parse_nbt -> decode_nbt

then 

decode_b64(string) -> string
encode_b64(string) -> string
decode_json(string) -> value
encode_json(value) -> string

### PR DESCRIPTION
**Pull request by Ghoulboy78**
Fixes #811 
Todo:

- [x] Testing properly i.e actually checking if this works. I has no reason not to as far as I can see, but I haven't run it yet
- [x] Docs

### CODE REVIEW

**Comment by altrisi** [comment]
Instead of a new function in ValueConversions you could just use `ScarpetJsonDeserializer`, or reuse the existing GSON instance which has JSON -> Value already prepared (I think it was somewhere around FileModule, but that has changed recently.

**Comment by Ghoulboy78** [comment]
That sounds like a very good idea, Imma check it out.

File: src/main/java/carpet/script/language/ControlFlow.java [hunk 1]
```diff
@@ -145,5 +151,23 @@ public static void apply(Expression expression) // public just to get the javado
                 return (c_, t_) -> retval;
             }
         });
+
+        expression.addUnaryFunction("encode_b64", v -> StringValue.of(Base64.getEncoder().encodeToString(v.getString().getBytes())));
+        expression.addUnaryFunction("decode_b64", v -> {
+            try {
+                return StringValue.of(new String(Base64.getDecoder().decode(v.getString()), StandardCharsets.ISO_8859_1));//using this charset cos it's the one used in decoding function
+            } catch (IllegalArgumentException iae){
+                throw new InternalExpressionException("Invalid b64 string: " + v.getString());
```
Comment (gnembon) [inline src/main/java/carpet/script/language/ControlFlow.java:160]: throw a user-catchable error rather here, there is already a json_error, just add base64_error

Comment (Ghoulboy78) [reply src/main/java/carpet/script/language/ControlFlow.java:160]: wdym user catchable?

File: src/main/java/carpet/script/language/ControlFlow.java [hunk 2]
```diff
@@ -145,5 +151,23 @@ public static void apply(Expression expression) // public just to get the javado
                 return (c_, t_) -> retval;
             }
         });
+
+        expression.addUnaryFunction("encode_b64", v -> StringValue.of(Base64.getEncoder().encodeToString(v.getString().getBytes())));
```
Comment (gnembon) [inline src/main/java/carpet/script/language/ControlFlow.java:155]: ControlFlow is for controlling flow of programs not for IO utils. Use Sys or DataStructures

File: src/main/java/carpet/script/language/ControlFlow.java [hunk 3]
```diff
@@ -145,5 +151,23 @@ public static void apply(Expression expression) // public just to get the javado
...
+        expression.addUnaryFunction("decode_b64", v -> {
+            try {
+                return StringValue.of(new String(Base64.getDecoder().decode(v.getString()), StandardCharsets.ISO_8859_1));//using this charset cos it's the one used in decoding function
```
Comment (gnembon) [inline src/main/java/carpet/script/language/ControlFlow.java:158]: Why not UTF-8?

File: src/main/java/carpet/script/language/ControlFlow.java (hunk 1 above)
Comment (altrisi) [reply src/main/java/carpet/script/language/ControlFlow.java:160]: ThrowStatement, from #692 + an edit I can't find the number rn. You have examples there.

(aka exceptions that can be caught by the user in a `try` block)

File: src/main/java/carpet/script/language/ControlFlow.java (hunk 2 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/language/ControlFlow.java:155]: k

File: src/main/java/carpet/script/language/ControlFlow.java (hunk 3 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/language/ControlFlow.java:158]: as I mentioned in the comment. that's the charset used in the `Base64.getDecoder.decode(String s)` function, so it makes sense to re-encode with the same charset.

File: src/main/java/carpet/script/language/ControlFlow.java (hunk 1 above)
Comment (Ghoulboy78) [reply src/main/java/carpet/script/language/ControlFlow.java:160]: oh ok

File: src/main/java/carpet/script/language/ControlFlow.java (hunk 3 above)
Comment (gnembon) [reply src/main/java/carpet/script/language/ControlFlow.java:158]: well entire scarpet base is in UTF-8 and you are potentially introducing a string with a different base

Comment (gnembon) [reply src/main/java/carpet/script/language/ControlFlow.java:158]: kk, gotcha

Comment (gnembon) [reply src/main/java/carpet/script/language/ControlFlow.java:158]: but you might need to test with some wierd emojis and stuff

**Comment by gnembon** [comment]
> ScarpetJsonDeserializer

Yeah, if you move the GSON static to ScarpetJsonDeserializer and use it in both places

**Comment by Ghoulboy78** [comment]
Ok this pr is ready to merge as well, I've tested and it works fine @gnembon

**Comment by altrisi** [comment]
Mind adding the exceptions that are thrown to the docs of each function, and to the list of exceptions?

**Comment by Ghoulboy78** [comment]
Sure thing mate

**Comment by Ghoulboy78** [comment]
done

File: docs/scarpet/language/LoopsAndHigherOrderFunctions.md [hunk 4]
```diff
@@ -146,3 +146,12 @@ it will be used from now on as a new value for the accumulator.
 reduce([1,2,3,4],_a+_,0)  => 10
 reduce([1,2,3,4],_a*_,1)  => 24
 </pre>
```
Comment (gnembon) [inline docs/scarpet/language/LoopsAndHigherOrderFunctions.md:148]: Why in loops 🤦

**Pull request merged**